File: best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int sum = 0;
        for (int i = 0; i < prices.length - 1; i++) {
            int diff = prices[i+1] - prices[i];
            if (sum < 0) {
                sum = diff;
            } else {
                sum += diff;
            }
            
            if (sum > profit) {
                profit = sum;
            }
        }
        
        return profit;
    }
}

File: best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        currMin = float('inf')
        
        profit = float('-inf')
        for i in range(len(prices)):
            if profit < prices[i] - currMin:
                profit = prices[i] - currMin

            currMin = min(currMin, prices[i])

        return profit if profit > 0 else 0
